### DISCUSSION TITLE
Installing MGIS for FEniCS

### DISCUSSION THREAD

**Discussion by mehranqsb** [thelfer/tfel#438] [Q&A]
Dear TFEL community,

I have a problem with installing mgis.fenics Python module. I downloaded MFrontGenericInterfaceSupport from [MFrontGenericInterfaceSupport](https://github.com/thelfer/MFrontGenericInterfaceSupport), then according to what is stated for Installing TFEL/MFront on [TFEL/MFront](https://tfel.sourceforge.net/install.html#sec:QuickUbuntu) following the Quick installation procedure on Ubuntu. I enabled bindings options for python and fenics, then after double configuting and generating. I tried to make install it, but this error occured during the installation process 

**"/home/mehran/TFEL/MFrontGenericInterfaceSupport-master/include/MGIS/ThreadedTaskResult.hxx:90:10: error: ‘optional’ in namespace ‘std’ does not name a template type
   90 |     std::optional<T> result;
"**

and 

**"/home/mehran/TFEL/MFrontGenericInterfaceSupport-master/include/MGIS/ThreadedTaskResult.ixx:29:9: error: class ‘mgis::ThreadedTaskResult<T>’ does not have any field named ‘result’
   29 |       : result(args...) {}  // end of ThreadedTaskResult
      |         ^~~~~~**

Could you please help me to solve this problem and accomplish the installation process.

Thanks in advance,
Best reagrds,
Mehran Ghasabeh

**Comment by thelfer** [comment]
Normally, the <optional> header is included at the beginning of this file. Your error may be caused by an old compiler version which only partially supports `C++17`.
Would you give the name and version of you compiler ?

**Reply by mehranqsb** [reply]
gcc (Ubuntu 10.5.0-1ubuntu1~22.04) 10.5.0

**Reply by mehranqsb** [reply]
I tried again now I got this error at the end of the installation 

Built target py_mgis_model
Install the project...
-- Install configuration: ""
-- Installing: /usr/local/share/doc/mgis/web/css
CMake Error at docs/web/cmake_install.cmake:46 (file):
  file INSTALL cannot make directory "/usr/local/share/doc/mgis/web/css": No
  such file or directory.
Call Stack (most recent call first):
  docs/cmake_install.cmake:47 (include)
  cmake_install.cmake:47 (include)

**Reply by thelfer** [reply]
This is because you can't install mgis at this location as a regular user. You can either:

- install as `root`. This is *not* recommended
- install in a local directory. You must add `-DCMAKE_INSTALL_PREVIX=...` to indicate a local directory. For instance, you can choose `$HOME/codes/mgis/master/install`. In this case, you must update your environment variables. I personnally use this script that I place at the root of the install directory (you must change the first line):

  ```sh
  export MGISHOME=/home/th202608/codes/mgis/master/install-python-3.7
  export LD_LIBRARY_PATH=$MGISHOME/lib:$LD_LIBRARY_PATH
  export PYTHONPATH=$MGISHOME/lib/python3.7/site-packages/:$PYTHONPATH
  ```
  Then everytime I use `mgis`, I do:
  
  ```sh
  $ source $HOME/codes/mgis/master/install-python-3.7/env.sh
  ```

**Reply by mehranqsb** [reply]
Thanks for your information and help, 
After cloning [MFrontGenericInterfaceSupport](https://github.com/thelfer/MFrontGenericInterfaceSupport).  I created 
a build-mgis directory under another directory named TFEL. Then I changed the directory to build-mgi and did 
ccmake ../MFrontGenericInterfaceSupport, then  
I set CMAKE_INSTALL_PREFIX  as  /home/mehran/TFEL/build-mgis/install, and only enabled python bindings (not fenics bindings). Then, after double configuring and generating, I did make && make install. It was accomplished without any error or warning. Then I added these lines to ~/.bashrc 

`export MGISHOME=/home/mehran/TFEL/build-mgis/install-python-3.10`
`export LD_LIBRARY_PATH=$MGISHOME/lib:$LD_LIBRARY_PATH`
`export PYTHONPATH=$MGISHOME/lib/python3.10/site-packages/:$PYTHONPATH`

Then, I executed bash on the terminal. In the end, I did 

`source $HOME/TFEL/build-mgis/install-python-3.10/env.sh`

But I got this message on the terminal.

`bash: /home/mehran/TFEL/build-mgis/install-python-3.10/env.sh: No such file or directory`

Best regards,
Mehran Ghasabeh

**Reply by thelfer** [reply]
Compare the two paths:

> CMAKE_INSTALL_PREFIX as /home/mehran/TFEL/build-mgis/install
> /home/mehran/TFEL/build-mgis/install-python-3.10/env.sh

Seems that the `-python-3.10` shall be removed

**Reply by mehranqsb** [reply]
But env.sh still cannot be found.

`bash: /home/mehran/TFEL/build-mgis/install/env.sh: No such file or directory`

**Reply by thelfer** [reply]
did you call `make` and `make install` ?

**Reply by mehranqsb** [reply]
I did them directly after configuring and generating the ccmake ../../MFrontGenericInterfaceSupport, inside the /home/mehran/TFEL/build-mgis directory.

**Reply by thelfer** [reply]
You mean that the install directory is also the build directory ? Please don't do that and use two distincts directories

**Reply by mehranqsb** [reply]
Alright! I will repeat the step again and define the different directory for installation.

**Reply by mehranqsb** [reply]
I repeated the above steps again, and mgis was installed. When I import mgis in python it works, but when I am running the docker to use the FEniCs project and call mgis, the module is not found (ModuleNotFoundError: No module named 'mgis').

**Reply by thelfer** [reply]
Did you check if the environment variables are correctly set ?

**Reply by mehranqsb** [reply]
These are my environment variables. I called them inside ~/.bashrc, and I executed the bash in the terminal before using FEniCs

`export MGISHOME=/home/mehran/TFEL/install`
`export LD_LIBRARY_PATH=$MGISHOME/lib:$LD_LIBRARY_PATH`
`export PYTHONPATH=$MGISHOME/lib/python3.10/site-packages/:$PYTHONPATH`

I want to note that my install directory is different than the build directory. 
I checked all the above directories and mgis is found in ''$MGISHOME/lib/python3.10/site-packages/''
